Inductor qlinear int8_fp32 with bmm

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116604
* __->__ #116599
* #116937


**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/116492, `linear` will be decomposed into `bmm` when input dim exceeds 2 and not contiguous. Fix this issue by convert the pattern back into `qlinear`. This PR focus on int8_fp32 case, will follow up int8_bf16 case in next PR.

**Test Plan**
```
python -u -m pytest -s -v test_mkldnn_pattern_matcher.py -k test_qlinear_input_dim_exceeds_2_and_not_contiguous
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
Inductor qlinear int8_fp32 with bmm

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
cc @jiayisunx @jgong5 please kindly help to take a look of this PR.

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"



**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/116492, linear will be decomposed into bmm when input dim exceeds 2 and not contiguous. Fix this issue by convert the pattern back into qlinear.

**Test Plan**
```
python -u -m pytest -s -v test_mkldnn_pattern_matcher.py -k test_qlinear_input_dim_exceeds_2_and_not_contiguous
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"



**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/116492, `linear` will be decomposed into `bmm` when input dim exceeds 2 and not contiguous. Fix this issue by convert the pattern back into `qlinear`. This PR focus on int8_fp32 case, will follow up int8_bf16 case in next PR.

**Test Plan**
```
python -u -m pytest -s -v test_mkldnn_pattern_matcher.py -k test_qlinear_input_dim_exceeds_2_and_not_contiguous
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -1344,38 +1360,80 @@ def _generate_qconv_weight_prepack_patterns(dtype=torch.float32):
-                    0
-                ]  # pattern: linear -> to_fb16 -> mul
-            )
+            # bmm pattern decomposed from linear when input dim exceeds 2 and not contiguous
```
add comment for the pattern here just like other paths?

leslie-fang-intel replied:
Thanks and add the comments for the linear patterns to do weight prepack.

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"



**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/116492, `linear` will be decomposed into `bmm` when input dim exceeds 2 and not contiguous. Fix this issue by convert the pattern back into `qlinear`. This PR focus on int8_fp32 case, will follow up int8_bf16 case in next PR.

**Test Plan**
```
python -u -m pytest -s -v test_mkldnn_pattern_matcher.py -k test_qlinear_input_dim_exceeds_2_and_not_contiguous
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
Hi @atalman, this PR failed to land. looks like we still need all files in PR to match one merge rule?

atalman commented:
@leslie-fang-intel and @jgong5 can we merge both groups ``x86 CPU quantization`` and ``CPU inductor`` these looks like closely related ?

leslie-fang-intel commented:
> @leslie-fang-intel and @jgong5 can we merge both groups x86 CPU quantization and CPU inductor these looks like closely related ?

I am ok with it. The only concern is that some of the files under `x86 CPU quantization` like `aten/src/ATen/native/quantized/cpu/**`, `test/quantization/core/test_quantized_op.py` are inductor irreverent.

Another way just comes to my mind is maybe we can move `torch/_inductor/fx_passes/quantization.py` from `x86 CPU quantization` into `CPU inductor`, it should cover most of the PRs.

- Since in general, we will change `torch/_inductor/fx_passes/quantization.py` and `test/inductor/test_mkldnn_pattern_matcher.py` together in one Inductor lowering specific PR.
- The remaining files under `x86 CPU quantization` label are op and recipe specific and in general cases we can put them in separate PRs.

@jgong5 what do you think about it?

jgong5 commented:
> @jgong5 what do you think about it?

I'm open to merge these two modules into one in the merge rule.

atalman commented:
@leslie-fang-intel and @jgong5 Sounds good, please post a PR to merge these rules. And tag me for approval. This should unblock this PR

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"



**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/116492, `linear` will be decomposed into `bmm` when input dim exceeds 2 and not contiguous. Fix this issue by convert the pattern back into `qlinear`. This PR focus on int8_fp32 case, will follow up int8_bf16 case in next PR.

**Test Plan**
```
python -u -m pytest -s -v test_mkldnn_pattern_matcher.py -k test_qlinear_input_dim_exceeds_2_and_not_contiguous
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Inductor qlinear int8_fp32 with bmm"



**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/116492, `linear` will be decomposed into `bmm` when input dim exceeds 2 and not contiguous. Fix this issue by convert the pattern back into `qlinear`. This PR focus on int8_fp32 case, will follow up int8_bf16 case in next PR.

**Test Plan**
```
python -u -m pytest -s -v test_mkldnn_pattern_matcher.py -k test_qlinear_input_dim_exceeds_2_and_not_contiguous
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
> groups `x86 CPU quantization` and `CPU inductor`

Thanks @atalman, https://github.com/pytorch/pytorch/pull/116937 opened for it.

leslie-fang-intel commented:
> @leslie-fang-intel and @jgong5 Sounds good, please post a PR to merge these rules. And tag me for approval. This should unblock this PR

Thanks @atalman, https://github.com/pytorch/pytorch/pull/116937 opened for it.

malfet commented:
Hmm, why this PR depends on https://github.com/pytorch/pytorch/pull/116937 ? (They are completely unrelated, aren't they?)

leslie-fang-intel commented:
> Hmm, why this PR depends on #116937 ? (They are completely unrelated, aren't they?)

Hi @malfet, this PR has changed 2 files which cross 2 different merge groups as `x86 CPU quantization` and `CPU inductor`. Per the discussion with @atalman above, due to the limitation of the current merging rules that prevent cross-checking all files among different merge groups, we can't merge it. So, we open [#116937](https://github.com/pytorch/pytorch/pull/116937) to workaround this issue per the discussion above.

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
